### PR TITLE
ENH: analytical formula for f distribution entropy

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
 gh-17748

#### Additional information
Kudos to @WarrenWeckesser for fixing the wrong formula found in many references such as wikipedia

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
Nice 👍 Is there a ref we can link to? If there are frequent issues, this deserves a comment so future maintainers don't "fix" this back.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:2737]: I notice that the mpmath reference here matches the SciPy implementation quite closely, substituting special functions from mpmath for those in SciPy. There are some cases in which this makes a good test, but it depends on the intent of the test. Since there are many of this form recently, I thought I'd check in about it.

In case the intent is to match the style of other tests, note that in many cases the formulation used in the reference mpmath implementation is different - e.g. using `quad`, the inverse of the function being tested, or its complement (like `1 - exp` in the mpmath reference instead of `expm1` in the SciPy implementation). These are often useful because they take advantage of mpmath's extra precision to allow for simpler implementations that can be compared directly against a reference formula or SciPy's expression of other distribution functions.

Or sometimes the implementation is very similar, but the tests check extreme values of parameters to make sure that the expression used in SciPy doesn't break down numerically. 

Is this one testing something like that? If it's mostly checking the correctness of SciPy's special functions (and NumPy arithmetic) in ordinary conditions, I think those are (or should be) be tested separately. In this case, perhaps checking against `mpmath.quad` with a simple expression of the PDF would be stronger because it would also show that this is indeed the entropy of the intended distribution (i.e. there is no mistake in the formula. In adds value to SciPy's generic distribution tests both because it's checking at more values of the parameters and because `mpmath.quad` can be more precise than `integrate.quad`.)

Comment (WarrenWeckesser) [reply scipy/stats/tests/test_distributions.py:2737]: When using mpmath to check results and the mpmath code is more or less a translation of the scipy code, there is definitely a concern that the translation includes the same mistakes as the scipy code, and all we've done is compute a more precise incorrect answer.  In this case, we can alleviate that concern by providing in the comments another implementation with mpmath, based on the integral definition of the differential entropy, that gives the same results, as you suggested.  For example, the function `f_entropy_with_quad(dfn, dfd)` here:

```
def f_entropy_with_quad(dfn, dfd):

    def pdf(x):
        # A direct translation of the PDF given in
        # https://en.wikipedia.org/wiki/F-distribution,
        # with d1 -> dfn and d2 -> dfd
        return (mp.sqrt((dfn*x)**dfn * dfd**dfd /
                        (dfn*x + dfd)**(dfn + dfd)) /
                (x*mp.beta(dfn/2, dfd/2)))

    return -mp.quad(lambda t: pdf(t)*mp.log(pdf(t)), [0, mp.inf])
```
with `mp.dps = 50`, gives float results that are the same as those already in the unit test.  This snippet could be added to the comment so that future readers can be confident (and can check themselves by running either version) that the expected values are the true expected 64 bit floating point values.

**Comment by dschmitz89** [comment]
I added the mpmath integration into the comments. Before merging, let's add @WarrenWeckesser as co-comitter.

**Comment by tupui** [comment]
@dschmitz89 you can do this if you re-write the commit message to:

```
TST: derive reference values from actual PDF integration

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>
```

p.s. try not to forget the 3 letters telling the commit's category.

**Comment by dschmitz89** [comment]
scipy keeps me on top of my git game indeed ;). Hope all correct now.

**Comment by tupui** [comment]
I think it did not work because it needs to be on it's own line. And also it should have the hyphen and colon. Yeah GitHub... 😅

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py]: (@dschmitz89 I thought we should discuss the following here rather than in gh-17774 so as not to distract the author.)

> `test_continuous_basic` does not actually test the computed entropy values against a reference. In case the `_entropy` method is overriden, it compares this result with the numerical integration in double precision of the parent class from what I see. 

This is correct: [`check_private_entropy`](https://github.com/scipy/scipy/blob/695ebde90f7374fdf35f75db940b90cd7ed28637/scipy/stats/tests/common_tests.py#L95-L99) compares entropy overrides against the generic implementation.

> This result can be also completely wrong unfortunately.

This is also true. But in the comment, this is used as motivation for why we need to check the results against an arbitrary precision library like mpmath. I don't think that conclusion is supported.

The fact that the generic implementation can be wrong is only likely to cause _false positives_ - test failures that do not actually indicate a problem with the overridden entropy method (just the inaccuracy of the generic implementation).

If the test _passes_, indicating strong agreement between the generic implementation agrees and the overridden implementation, I would be very surprised if this were due to both implementations being wrong yet producing the same result (especially if the result is not some trivial number like 0). It seems more likely to me that these (presumably very different) implementations are both right - or at least they are the entropy of the PDF that we implement.

For this reason, I'd suggest that no distribution-specific entropy tests are necessary unless we can identify a reason that the generic test is insufficient. One example is if the overridden entropy implementation is specially written to be accurate for outrageous or tricky values of the distribution parameters. Another reason would be if the implementation of the entropy branches depending on the values of the parameters but not all of these branches are exercised by the generic test (because they are not yet included in `_distr_params.py`.

This would make review of gh-17774 easier, for instance. Rather than asking the author to rewrite the tests he added for you (so that the tests check that the entropy is correct rather than checking the correctness of array arithmetic and SciPy special functions), the tests could simply be removed.

What do you think?

Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py]: This actually made me think and I tend to agree with you. Both implementations wrong at the same time would be a very rare occasion.

For problematic entropy formulas like here and some of #17774 , we should add a comment in the code though where it comes from and compare to mpmath test cases in the PR thread which will not be added to the test suite.

**Comment by tupui** [comment]
@dschmitz89 seems like there was an issue with you last push. The test file appears to be empty for GitHub.

**Comment by dschmitz89** [comment]
> @dschmitz89 seems like there was an issue with you last push. The test file appears to be empty for GitHub.

I think this is github's weird way of showing that a file was first touched in a PR and that then the change was reverted. The test file was definitely not deleted. New tests were not added as the analytical formula does unfortunately not improve precision (for small values mpmath gives different results). The formula is as reliable as the current numerical integration but much faster.

**Review by tupui — APPROVED** [review]
Thanks for the precision, I can see that selecting one commit at a time 👍 Then LGTM and unless I am misreading seems other maintainers are ok.

I will squash merge and add Warren.

**Pull request merged**
